### PR TITLE
providing diagnostic error report for I/O related errors

### PR DESCRIPTION
**Pull request by Jarvi-Izana**
The motivation of this is to provide better diagnostic information for s3 related issues. Later I realize that it's plausible to apply for all I/O related issues.

### CODE REVIEW

File: src/core/storage/fileio/s3_api.cpp [hunk 1]
```diff
@@ -363,11 +382,15 @@ list_objects_response list_objects_impl(s3url parsed_url,
...
-                    outcome.GetError().GetExceptionName() <<
-                    ", msg: " << outcome.GetError().GetMessage() << std::endl;
-            ret.error = stream.str();
+            std::stringstream ss;
+            ss << '(' << parsed_url
+               << ".proxy:" << clientConfiguration.proxyHost << ".region:"
+               << clientConfiguration.region << ')'
+               << " Error while listing Objects, exception: "
+               << outcome.GetError().GetExceptionName()
+               << ".Error msg from aws: " << outcome.GetError().GetMessage()
```
Comment (TobyRoseman) [inline src/core/storage/fileio/s3_api.cpp]: `aws` -> `AWS`

Comment (TobyRoseman) [reply src/core/storage/fileio/s3_api.cpp]: same comment else where you use `aws`.

File: src/core/storage/fileio/s3_api.cpp [hunk 2]
```diff
@@ -363,11 +382,15 @@ list_objects_response list_objects_impl(s3url parsed_url,
...
         else
         {
-            std::stringstream stream;
-            stream << "Error while listing Objects, exception: " <<
-                    outcome.GetError().GetExceptionName() <<
-                    ", msg: " << outcome.GetError().GetMessage() << std::endl;
-            ret.error = stream.str();
+            std::stringstream ss;
+            ss << '(' << parsed_url
+               << ".proxy:" << clientConfiguration.proxyHost << ".region:"
```
Comment (TobyRoseman) [inline src/core/storage/fileio/s3_api.cpp]: Do you want a space after the periods?

File: src/core/storage/fileio/fs_utils.cpp [hunk 3]
```diff
@@ -145,33 +146,32 @@ EXPORT file_status get_file_status(const std::string& path) {
       // get the HDFS object
       auto& hdfs = turi::hdfs::get_hdfs(host, std::stoi(port));
       // fail we we are unable to construct the HDFS object
-      if (!hdfs.good()) return file_status::FS_UNAVAILABLE;
+      if (!hdfs.good()) return std::make_pair(file_status::FS_UNAVAILABLE, "unable to construct through:" + path);
```
Comment (hoytak) [inline src/core/storage/fileio/fs_utils.cpp]: Needs space after ":".

File: src/core/storage/fileio/s3_api.hpp [hunk 4]
```diff
@@ -34,6 +36,12 @@ struct s3url {
            object_name == other.object_name &&
            endpoint == other.endpoint;
   }
+
+  friend std::ostream& operator<<(std::ostream& os, const s3url& url) {
+    os << "bucket:" << url.bucket << ", object_name:" << url.object_name
```
Comment (hoytak) [inline src/core/storage/fileio/s3_api.hpp]: spaces after the :

File: src/core/storage/fileio/s3_api.hpp [hunk 5]
```diff
@@ -204,8 +211,37 @@ void set_download_timeout(long timeout);
...
+  static const std::vector<std::string> _enum_to_str;
+};
 
+template <class Response>
+std::ostream& reportS3Error(std::ostream& ss, const s3url& parsed_url,
+                            S3Operation::ops_enum operation,
+                            const Aws::Client::ClientConfiguration& config,
+                            const Response& outcome) {
+  ss << '(' << parsed_url << ", proxy:" << config.proxyHost
+     << ", region:" << config.region << ')'
```
Comment (hoytak) [inline src/core/storage/fileio/s3_api.hpp]: spaces after the :, here and one line above.

**Comment by Jarvi-Izana** [comment]
Thanks @hoytak for the help! I forget to push the changes on my local machine.

**Comment by Jarvi-Izana** [comment]
Hi @hoytak, I just remember the initial design of not having spaces is due to the reason that I want to get users easier to realize the leading spaces of url may cause trouble. I had similar issues before and it took me a long time to realize that aws won't trim url. My proposal is to wrap url with single quotes if spaces are provided for the readability.

**Comment by Jarvi-Izana** [comment]
Ready to roll once the internal build passes.

**Comment by hoytak** [comment]
Looks like we just need to change the type of exception looked for in the test SFrameTest.test_creation_from_csv_dir_local of test_sframe from OSError to IOError.

**Comment by Jarvi-Izana** [comment]
That's right. It's included in my last commit. Wait for your @hoytak fix to the pipeline.

**Pull request merged**